ledger/sandbox-on-x/src/classic/scala/com/daml/platform/ledger/SandboxServer.scala

```diff
@@ -104,25 +117,65 @@ final class SandboxServer(
       .map(path => Future(Files.write(path, Seq(port.toString).asJava)).map(_ => ()))
       .getOrElse(Future.unit)
 
-  private def loadPackages(writeService: WriteService)(implicit
-      executionContext: ExecutionContext
-  ): AbstractResourceOwner[ResourceContext, List[Unit]] =
-    ResourceOwner.forFuture(() =>
-      Future.sequence(
-        config.damlPackages.map { file =>
-          val submissionId = Ref.SubmissionId.assertFromString(UUID.randomUUID().toString)
-          for {
-            dar <- Future.fromTry(DarParser.readArchiveFromFile(file).toTry)
-            _ <- writeService
-              .uploadPackages(submissionId, dar.all, None)(
-                LoggingContext.ForTesting,
-                NoOpTelemetryContext,
+  private def loadPackages(writeService: WriteService, indexService: IndexService)(implicit
+      executionContext: ExecutionContext,
+      system: ActorSystem,
+      loggingContext: LoggingContext,
+  ): AbstractResourceOwner[ResourceContext, Unit] =
+    ResourceOwner.forFuture(() => {
+      val packageUploadTimeout = 10.seconds
+
+      val packageSubmissionsTrackerMap = config.damlPackages.map { file =>
+        val uploadCompletionPromise = Promise[Unit]()
+        system.scheduler.scheduleOnce(packageUploadTimeout) {
+          // Package upload timeout, meant to allow fail-fast for testing
+          // TODO Remove once in-memory backend (e.g. H2) is deemed stable
+          uploadCompletionPromise.tryFailure(
+            new RuntimeException(s"Package upload timeout after $packageUploadTimeout")
+          )
+          ()
+        }
+        UUID.randomUUID().toString -> (file, uploadCompletionPromise)
+      }.toMap
+
+      val (killSwitch, packageEntriesStreamDone) = indexService
+        .packageEntries(None)
+        .viaMat(KillSwitches.single)(Keep.right[NotUsed, UniqueKillSwitch])
+        .toMat(Sink.foreach {
+          case PackageEntry.PackageUploadAccepted(submissionId, _) =>
+            packageSubmissionsTrackerMap(submissionId)._2.complete(Success(()))
```

Comment:
`packageSubmissionsTrackerMap(submissionId)` would most probably return no value as there is random UUID assigned there and not a `submissionId`, am I right?

Reply:
actually I found out how it works, nvm

Reply:
The random uuid is the submission id of the package upload request. See 17 lines below 
```
              writeService
                .uploadPackages(Ref.SubmissionId.assertFromString(submissionId), dar.all, None)(
                  LoggingContext.ForTesting,
                  NoOpTelemetryContext,
                )
```